Two process instances with different process definitions share the same key

**Describe the bug**

While investigating an issue on the Operate, we came across a case where two process instances, with two different process definitions, share the same key. Not only is the process instance key the same, the record key is the same. However, the positions of the records are different.

Here are the records pulled from Operate:

```json
{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 3,
    "successful" : 3,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 2,
      "relation" : "eq"
    },
    "max_score" : 4.890227,
    "hits" : [
      {
        "_index" : "zeebe-record_process-instance_1.2.3_2021-11-02",
        "_type" : "_doc",
        "_id" : "48-1456937",
        "_score" : 4.890227,
        "_routing" : "48",
        "_source" : {
          "partitionId" : 48,
          "value" : {
            "version" : 2,
            "bpmnElementType" : "PROCESS",
            "bpmnProcessId" : "OrderOrchestrationNew",
            "processDefinitionKey" : 2251799813731929,
            "elementId" : "OrderOrchestrationNew",
            "processInstanceKey" : 108086391057176360,
            "flowScopeKey" : -1,
            "parentProcessInstanceKey" : -1,
            "parentElementInstanceKey" : -1
          },
          "key" : 108086391057176360,
          "sourceRecordPosition" : 1456910,
          "timestamp" : 1635849318977,
          "valueType" : "PROCESS_INSTANCE",
          "recordType" : "EVENT",
          "intent" : "ELEMENT_COMPLETED",
          "rejectionType" : "NULL_VAL",
          "rejectionReason" : "",
          "brokerVersion" : "1.2.3",
          "position" : 1456937
        }
      },
      {
        "_index" : "zeebe-record_process-instance_1.2.3_2021-11-02",
        "_type" : "_doc",
        "_id" : "48-1621008",
        "_score" : 4.890227,
        "_routing" : "48",
        "_source" : {
          "partitionId" : 48,
          "value" : {
            "version" : 2,
            "bpmnElementType" : "PROCESS",
            "bpmnProcessId" : "SingleLineNew",
            "processDefinitionKey" : 2251799813731931,
            "flowScopeKey" : -1,
            "parentProcessInstanceKey" : -1,
            "parentElementInstanceKey" : -1,
            "elementId" : "SingleLineNew",
            "processInstanceKey" : 108086391057176360
          },
          "key" : 108086391057176360,
          "sourceRecordPosition" : 1620977,
          "position" : 1621008,
          "timestamp" : 1635855215725,
          "valueType" : "PROCESS_INSTANCE",
          "recordType" : "EVENT",
          "intent" : "ELEMENT_COMPLETED",
          "rejectionType" : "NULL_VAL",
          "rejectionReason" : "",
          "brokerVersion" : "1.2.3"
        }
      }
    ]
  }
}
```

**To Reproduce**

Unclear

**Expected behavior**

Every entity should have a unique key! Keys should not be reused for different entities.


**Environment:**
 - OS: Camunda Cloud CCSaaS
 - Zeebe Version: 1.2.3
 - Configuration: Camunda Cloud CCSaaS

I can reproduce it with an IT Test:

```java
 @Test
  public void shouldNotProduceDuplicatedKeys() {
    // given
    final var previousLeaderId = clusteringRule.getLeaderForPartition(1).getNodeId();
    final var previousLeader = clusteringRule.getBroker(previousLeaderId);
    client.newDeployCommand().addProcessModel(PROCESS, PROCESS_RESOURCE_NAME).send().join();
    client
        .newCreateInstanceCommand()
        .bpmnProcessId("process")
        .latestVersion()
        .withResult()
        .send()
        .join();

    // disconnect leader - becomes follower
    clusteringRule.disconnect(previousLeader);
    clusteringRule.stepDown(previousLeader, 1);
    // remove old log content
    RecordingExporter.reset();
    final var newLeaderInfo = clusteringRule.awaitOtherLeader(1, previousLeaderId);
    final var newLeaderId = newLeaderInfo.getNodeId();
    assertThat(newLeaderId).isNotEqualTo(previousLeaderId);
    final var newLeader = clusteringRule.getBroker(newLeaderId);

    // old log is replayed on new leader so we can collect keys
    final var previousKeys =
        RecordingExporter.processInstanceRecords()
            .limitToProcessInstanceCompleted()
            .collect(Collectors.toList())
            .stream()
            .map(Record::getKey)
            .collect(Collectors.toList());

    // remove all content again
    RecordingExporter.reset();

    // produce new stuff on new leader
    client
        .newCreateInstanceCommand()
        .bpmnProcessId("process")
        .latestVersion()
        .withResult()
        .send()
        .join();

    final var newKeys =
        RecordingExporter.processInstanceRecords()
            .limitToProcessInstanceCompleted()
            .collect(Collectors.toList())
            .stream()
            .map(Record::getKey)
            .collect(Collectors.toList());

    assertThat(previousKeys).doesNotContain(newKeys.toArray(new Long[0]));
  }
```

Output:

```

java.lang.AssertionError: 
Expecting
  [2251799813685251L,
    2251799813685251L,
    2251799813685251L,
    -1L,
    2251799813685253L,
    2251799813685253L,
    2251799813685253L,
    2251799813685253L,
    2251799813685253L,
    2251799813685254L,
    2251799813685255L,
    2251799813685255L,
    2251799813685255L,
    2251799813685255L,
    2251799813685255L,
    2251799813685251L,
    2251799813685251L,
    2251799813685251L]
not to contain
  [2251799813685252L,
    2251799813685252L,
    2251799813685252L,
    -1L,
    2251799813685254L,
    2251799813685254L,
    2251799813685254L,
    2251799813685254L,
    2251799813685254L,
    2251799813685255L,
    2251799813685256L,
    2251799813685256L,
    2251799813685256L,
    2251799813685256L,
    2251799813685256L,
    2251799813685252L,
    2251799813685252L,
    2251799813685252L]
but found
  [-1L, 2251799813685254L, 2251799813685255L]

```



It is exactly how it was described by @saig0, if the follower has nothing to replay it will not set the key in the state correctly which completely breaks the keygenerator, since it starts at the beginning or on the last persisted state.


Compact Logger output:

```
23:37:31.572 [] [main] INFO  io.camunda.zeebe.test - Compact log representation:
--------
	['C'ommand/'E'event/'R'ejection] [valueType] [intent] - #[position]->#[source record position]  P[partitionId]K[key] - [summary of value]
	P9K999 - key; #999 - record position; "ID" element/process id; @"elementid"/[P9K999] - element with ID and key
	Keys are decomposed into partition id and per partition key (e.g. 2251799813685253 -> P1K005). If single partition, the partition is omitted.
	Long IDs are shortened (e.g. 'startEvent_5d56488e-0570-416c-ba2d-36d2a3acea78' -> 'star..acea78'
--------
C PROC_INST ACTIVATE       - #06->#05 K03 - PROCESS "process" in <process "process"[K03]>
E PROC_INST ACTIVATING     - #08->#06 K03 - PROCESS "process" in <process "process"[K03]>
E PROC_INST ACTIVATED      - #09->#06 K03 - PROCESS "process" in <process "process"[K03]>
C PROC_INST ACTIVATE       - #10->#06  -1 - START_EVENT "startEv..6475ce1" in <process "process"[K03]>
E PROC_INST ACTIVATING     - #11->#10 K05 - START_EVENT "startEv..6475ce1" in <process "process"[K03]>
E PROC_INST ACTIVATED      - #12->#10 K05 - START_EVENT "startEv..6475ce1" in <process "process"[K03]>
C PROC_INST COMPLETE       - #13->#10 K05 - START_EVENT "startEv..6475ce1" in <process "process"[K03]>
E PROC_INST COMPLETING     - #14->#13 K05 - START_EVENT "startEv..6475ce1" in <process "process"[K03]>
E PROC_INST COMPLETED      - #15->#13 K05 - START_EVENT "startEv..6475ce1" in <process "process"[K03]>
E PROC_INST SEQ_FLOW_TAKEN - #16->#13 K06 - SEQUENCE_FLOW "sequenc..c7d5287" in <process "process"[K03]>
C PROC_INST ACTIVATE       - #17->#13 K07 - END_EVENT "endEven..b1b69b4" in <process "process"[K03]>
E PROC_INST ACTIVATING     - #18->#17 K07 - END_EVENT "endEven..b1b69b4" in <process "process"[K03]>
E PROC_INST ACTIVATED      - #19->#17 K07 - END_EVENT "endEven..b1b69b4" in <process "process"[K03]>
E PROC_INST COMPLETING     - #20->#17 K07 - END_EVENT "endEven..b1b69b4" in <process "process"[K03]>
E PROC_INST COMPLETED      - #21->#17 K07 - END_EVENT "endEven..b1b69b4" in <process "process"[K03]>
C PROC_INST COMPLETE       - #22->#17 K03 - PROCESS "process" in <process "process"[K03]>
E PROC_INST COMPLETING     - #23->#22 K03 - PROCESS "process" in <process "process"[K03]>
E PROC_INST COMPLETED      - #24->#22 K03 - PROCESS "process" in <process "process"[K03]>
C PROC_INST ACTIVATE       - #26->#25 K04 - PROCESS "process" in <process "process"[K04]>
E PROC_INST ACTIVATING     - #28->#26 K04 - PROCESS "process" in <process "process"[K04]>
E PROC_INST ACTIVATED      - #29->#26 K04 - PROCESS "process" in <process "process"[K04]>
C PROC_INST ACTIVATE       - #30->#26  -1 - START_EVENT "startEv..6475ce1" in <process "process"[K04]>
E PROC_INST ACTIVATING     - #31->#30 K06 - START_EVENT "startEv..6475ce1" in <process "process"[K04]>
E PROC_INST ACTIVATED      - #32->#30 K06 - START_EVENT "startEv..6475ce1" in <process "process"[K04]>
C PROC_INST COMPLETE       - #33->#30 K06 - START_EVENT "startEv..6475ce1" in <process "process"[K04]>
E PROC_INST COMPLETING     - #34->#33 K06 - START_EVENT "startEv..6475ce1" in <process "process"[K04]>
E PROC_INST COMPLETED      - #35->#33 K06 - START_EVENT "startEv..6475ce1" in <process "process"[K04]>
E PROC_INST SEQ_FLOW_TAKEN - #36->#33 K07 - SEQUENCE_FLOW "sequenc..c7d5287" in <process "process"[K04]>
C PROC_INST ACTIVATE       - #37->#33 K08 - END_EVENT "endEven..b1b69b4" in <process "process"[K04]>
E PROC_INST ACTIVATING     - #38->#37 K08 - END_EVENT "endEven..b1b69b4" in <process "process"[K04]>
E PROC_INST ACTIVATED      - #39->#37 K08 - END_EVENT "endEven..b1b69b4" in <process "process"[K04]>
E PROC_INST COMPLETING     - #40->#37 K08 - END_EVENT "endEven..b1b69b4" in <process "process"[K04]>
E PROC_INST COMPLETED      - #41->#37 K08 - END_EVENT "endEven..b1b69b4" in <process "process"[K04]>
C PROC_INST COMPLETE       - #42->#37 K04 - PROCESS "process" in <process "process"[K04]>
E PROC_INST COMPLETING     - #43->#42 K04 - PROCESS "process" in <process "process"[K04]>
E PROC_INST COMPLETED      - #44->#42 K04 - PROCESS "process" in <process "process"[K04]>

-------------- Deployed Processes ----------------------

--------------- Decomposed keys (for debugging) -----------------
 -1 <-> -1
K03 <-> 2251799813685251
K04 <-> 2251799813685252
K05 <-> 2251799813685253
K06 <-> 2251799813685254
K07 <-> 2251799813685255
K08 <-> 2251799813685256
```


Actually I was thinking whether this can cause also the issue which @pihme has investigated regarding inconsistent state, since we use the same key on different executions? https://github.com/camunda-cloud/zeebe/issues/7988#issuecomment-949425963

--- a/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java
@@ -72,7 +72,6 @@ public final class ReplayStateMachine implements LogRecordAwaiter {
   private long batchSourceEventPosition = StreamProcessor.UNSET_POSITION;
 
   private long snapshotPosition;
-  private long highestRecordKey = -1L;
   private long lastReadRecordPosition = StreamProcessor.UNSET_POSITION;
   private long lastReplayedEventPosition = StreamProcessor.UNSET_POSITION;
 
@@ -234,9 +233,6 @@ public final class ReplayStateMachine implements LogRecordAwaiter {
    * the last source event, which has caused the last applied event.
    */
   private void onRecordsReplayed() {
-    // restore the key generate with the highest key from the log
-    keyGeneratorControls.setKeyIfHigher(highestRecordKey);
-
     LOG.info(LOG_STMT_REPLAY_FINISHED, lastReadRecordPosition);
     recoveryFuture.complete(lastSourceEventPosition);
   }
@@ -267,8 +263,7 @@ public final class ReplayStateMachine implements LogRecordAwaiter {
 
     // records from other partitions should not influence the key generator of this partition
     if (Protocol.decodePartitionId(currentRecordKey) == zeebeState.getPartitionId()) {
-      // remember the highest key on the stream to restore the key generator after replay
-      highestRecordKey = Math.max(currentRecordKey, highestRecordKey);
+      keyGeneratorControls.setKeyIfHigher(currentRecordKey);
     }
   }
 

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/ContinuouslyReplayTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/ContinuouslyReplayTest.java
@@ -62,8 +62,6 @@ public class ContinuouslyReplayTest {
               processingState.entrySet().stream()
                   .filter(entry -> entry.getKey() != ZbColumnFamilies.DEFAULT)
                   // ignores transient states
-                  .filter(entry -> entry.getKey() != ZbColumnFamilies.KEY)
-                  // on followers we don't need to reset the key
                   // this will happen anyway then on leader replay
                   .forEach(
                       entry -> {

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/FailOverReplicationTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/FailOverReplicationTest.java
@@ -15,7 +15,9 @@ import io.camunda.zeebe.client.ZeebeClient;
 import io.camunda.zeebe.it.util.GrpcClientRule;
 import io.camunda.zeebe.model.bpmn.Bpmn;
 import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
+import io.camunda.zeebe.protocol.record.Record;
 import io.camunda.zeebe.snapshots.SnapshotId;
+import io.camunda.zeebe.test.util.record.RecordingExporter;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.nio.file.Files;
@@ -177,6 +179,65 @@ public class FailOverReplicationTest {
 
     assertThat(nextLeader.getNodeId()).isEqualTo(getNodeId(followerA));
     clusteringRule.waitForSnapshotAtBroker(previousLeader);
+  }
+
+  @Test
+  public void shouldNotProduceDuplicatedKeys() {
+    // given
+    final var previousLeaderId = clusteringRule.getLeaderForPartition(1).getNodeId();
+    final var previousLeader = clusteringRule.getBroker(previousLeaderId);
+    client.newDeployCommand().addProcessModel(PROCESS, PROCESS_RESOURCE_NAME).send().join();
+    client
+        .newCreateInstanceCommand()
+        .bpmnProcessId("process")
+        .latestVersion()
+        .withResult()
+        .send()
+        .join();
+
+    // disconnect leader - becomes follower
+    clusteringRule.disconnect(previousLeader);
+    clusteringRule.stepDown(previousLeader, 1);
+    // remove old log content
+    RecordingExporter.reset();
+    final var newLeaderInfo = clusteringRule.awaitOtherLeader(1, previousLeaderId);
+    final var newLeaderId = newLeaderInfo.getNodeId();
+    assertThat(newLeaderId).isNotEqualTo(previousLeaderId);
+
+    // old log is replayed on new leader so we can collect keys
+    final var previousKeys =
+        RecordingExporter.processInstanceRecords()
+            .limitToProcessInstanceCompleted()
+            .collect(Collectors.toList())
+            .stream()
+            .map(Record::getKey)
+            .filter(k -> k != -1L)
+            .collect(Collectors.toList());
+
+    // remove all content again
+    RecordingExporter.reset();
+
+    // produce new stuff on new leader
+    client
+        .newCreateInstanceCommand()
+        .bpmnProcessId("process")
+        .latestVersion()
+        .withResult()
+        .send()
+        .join();
+
+    final var newKeys =
+        RecordingExporter.processInstanceRecords()
+            .limitToProcessInstanceCompleted()
+            .collect(Collectors.toList())
+            .stream()
+            .map(Record::getKey)
+            .filter(k -> k != -1L)
+            .toArray(Long[]::new);
+
+    assertThat(previousKeys)
+        .describedAs("Keys should always be unique for different entities.")
+        .doesNotContain(newKeys);
   }
 
   private int getNodeId(final Broker broker) {

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/FailOverReplicationTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/FailOverReplicationTest.java
@@ -182,8 +182,10 @@ public class FailOverReplicationTest {
   }
 
   @Test
+  // regression test https://github.com/camunda-cloud/zeebe/issues/8129
   public void shouldNotProduceDuplicatedKeys() {
     // given
+    // we produce some records on the old leader
     final var previousLeaderId = clusteringRule.getLeaderForPartition(1).getNodeId();
     final var previousLeader = clusteringRule.getBroker(previousLeaderId);
     client.newDeployCommand().addProcessModel(PROCESS, PROCESS_RESOURCE_NAME).send().join();
@@ -195,11 +197,14 @@ public class FailOverReplicationTest {
         .send()
         .join();
 
-    // disconnect leader - becomes follower
+    // we disconnect the leader and step down
     clusteringRule.disconnect(previousLeader);
     clusteringRule.stepDown(previousLeader, 1);
-    // remove old log content
+    // remove old exported records content - for easier verification later
     RecordingExporter.reset();
+
+    // when
+    // we await a new leader
     final var newLeaderInfo = clusteringRule.awaitOtherLeader(1, previousLeaderId);
     final var newLeaderId = newLeaderInfo.getNodeId();
     assertThat(newLeaderId).isNotEqualTo(previousLeaderId);
@@ -235,8 +240,10 @@ public class FailOverReplicationTest {
             .filter(k -> k != -1L)
             .toArray(Long[]::new);
 
+    // then
     assertThat(newKeys).isNotEmpty();
     assertThat(previousKeys)
+        .isNotEmpty()
         .describedAs("Keys should always be unique for different entities.")
         .doesNotContain(newKeys);
   }

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/FailOverReplicationTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/FailOverReplicationTest.java
@@ -235,6 +235,7 @@ public class FailOverReplicationTest {
             .filter(k -> k != -1L)
             .toArray(Long[]::new);
 
+    assertThat(newKeys).isNotEmpty();
     assertThat(previousKeys)
         .describedAs("Keys should always be unique for different entities.")
         .doesNotContain(newKeys);

